docs/design/2023-09-22-tiflash-query-level-auto-spill.md

```diff
@@ -0,0 +1,55 @@
+
+* [OperatorSpillContext](https://github.com/pingcap/tiflash/blob/ce24115c147293df10a9d9f1c745eb0e81f0f7b5/dbms/src/Core/OperatorSpillContext.h#L33), it is a structure that contains information about how many memories an operator may released when spill, for each operator that support spill, it will hold an `OperatorSpillContext`
+* [TaskOperatorSpillContexts](https://github.com/pingcap/tiflash/blob/ce24115c147293df10a9d9f1c745eb0e81f0f7b5/dbms/src/Core/TaskOperatorSpillContexts.h#L21), it is a collection of all `OperatorSpillContext`s that belongs to a `MPPTask`
+* [QueryOperatorSpillContexts](https://github.com/pingcap/tiflash/blob/ce24115c147293df10a9d9f1c745eb0e81f0f7b5/dbms/src/Core/QueryOperatorSpillContexts.h#L23), it is a collection of all `TaskOperatorSpillContexts` that belongs to a query
+
+As we can see, `QueryOperatorSpillContexts` has all the information about how many memory each opeartor can release when spill, so in `QueryOperatorSpillContexts`, it is easy to find out the opeartor that takes up most memory and mark it to spill. In current implementation, `QueryOperatorSpillContexts` will first find out the `TaskOperatorSpillContexts` that takes up most memories, and the choosen `TaskOperatorSpillContexts` will trigger spill for the `OperatorSpillContext` that takes up most memories.
+
+### When to trigger auto spill
+
+In operator's spill, spill is triggered when opeartor's memory usage exceeds the threshold, it does not work for query level's spill, because the query level's memory usage threshold is a hard limit, that is to say if the memory usage exceeds the threshold, TiFlash will throw error and abort the query immediately. So for query level auto spill, it has to trigger auto spill check before the memory usage is too large. In the implementation, there is a new variable to control when to trigger auto spill check: [auto_memory_revoke_trigger_threshold](https://github.com/pingcap/tiflash/blob/ce24115c147293df10a9d9f1c745eb0e81f0f7b5/dbms/src/Interpreters/Settings.h#L317). If the memory usage exceeds `auto_memory_revoke_trigger_threshold * query_memory_usage_limit`, TiFlash will trigger auto spill check, and try to mark operator to spill.
+
+### How to trigger auto spill check
+
+In operator's spill, auto spill check is triggered after each calculation of a input block in each opeartor that support spill. For query level auto spill, the basic idea is almost the same: check auto spill after each calculation of a input block. However, the difference is for query level auto spill, each operator, even if it does not support spill, will consume memory during calculation, so unlike operator's spill, in order to support query level auto spill, auto spill check should be trigger in all operators. In order to avoid add trigger code inside each operator, the implementation add this trigger code to the runtime framework
```

Comment:
```suggestion
In operator's spill, auto spill check is triggered after each calculation of a input block in each opeartor that supports spill. For query level auto spill, the basic idea is almost the same: check auto spill after each calculation of a input block. However, the difference is for query level auto spill, each operator, even if it does not support spill, will consume memory during calculation, so unlike operator's spill, in order to support query level auto spill, auto spill check should be triggered in all operators. In order to avoid adding trigger code inside each operator, the implementation adds this trigger code to the runtime framework
```